### ISSUE TITLE
[RFC] Overhaul script processing model to allow pre-compilation

### ISSUE THREAD

**Issue by lihaoyi** [com-lihaoyi/Ammonite#1228]
# Problem

The current script-processing model is closely modelled after the interactive, dynamic REPL experience:

1. Scripts are split via `@` separators into blocks
2. The blocks of a script are evaluated one at a time, with the following steps:
  - Magic-import analysis and evaluation. For any `import $file`s, they are handled recursived via (1.)
  - Code block compilation
  - Code block execution
3. Scripts and imported scripts are compiled and run one by one, until all are evaluated. 
 
While it works, that is generally not what people want in scripts, and has resulted in some fundamental limitations:

- We cannot easily integrate with BSP, without going through an entirely different compilation code path
- We cannot pre-compile scripts without running them
- Compiling large import graphs of scripts is very slow, because each imported script is compiled individually rather than together in a batch
- Retrofitting incremental compilation via Zinc is not possible.
- No way of supporting e.g. Scala.js or Scala-Native

The fundamental issue at hand is that the dynamic REPL-like evaluation of scripts is not amenable to static analysis and processing, the kind that is necessary for IDE support via BSP, pre-compilation, or incremental compilation. While it allows some interesting abilities like an imported script setting compilation files for the importer to make use of, these abilities are generally not widely used.

# Proposal 

This proposal is to replace the current dynamic module evaluation model with a more static one:

1. Scripts are analyzed for magic imports. Any file imports are recursively handled via (1.), until the full set of scripts and full set of magic imports is known
2. All necessary magic imports are evaluated: dependencies downloaded, etc.
3. `$configureCompiler{...}` blocks are identified in the transitive set of scripts, and executed in arbitrary order.
4. All collected scripts are compiled together in one combined compilation unit
5. The compiled output is executed

This model is similar to how the BSP plugin works today, and will solve all the problems above: no more differing code paths between the IDE and real execution, ease of pre-compilation (just stop after step 4), batched compilation speeding things up, and it now becomes possible to retrofit support for Zinc, Scala.js, or Scala-Native.

There is definitely a loss of flexibility in this model, but given the many benefits it's probably a price worth paying. 

# Internals

Such a change would be a significant split in Ammonite's functionality, and probably deserve a corresponding split in the codebase. We should introduce a new `scripts/` submodule that depends on `interp/`, and move all scripting related functionality there. That will live in parallel with the `repl/` submodule that depends on `interp/` and avoid having unnecessary script-related functionality in the REPL classpath.

**Comment by lihaoyi** [comment]
@alexarchambault how hard do you think this would be? it seems like we already have most of this logic present in your BSP work, so if we agree on how we want Ammonite scripts to work to cut over to this new model shouldn't be to difficult

**Comment by alexarchambault** [comment]
@lihaoyi I agree with the general direction, and yes, BSP support mostly already follows that model. But, to be frank, the script runner you're looking for is basically… [Scala CLI](https://github.com/VirtusLab/scala-cli).

It already supports scripts, with [slightly different semantics](https://scala-cli.virtuslab.org/docs/guides/scripts#difference-with-ammonite-scripts), and has most of the features you're discussing, and many more:
- Scala.JS / Scala Native support (`scala-cli foo.sc --js`, `scala-cli foo.sc --native`)
- BSP / Metals support (only for `.scala` files with [today's instructions](https://scala-cli.virtuslab.org/docs/guides/ide), but support for `.sc` files has already been added [on a Metals fork](https://github.com/VirtuslabRnD/metals/tree/scala-cli-v1), re-using the Metals Ammonite support)
- packaging, even for Scala.JS / Scala Native (`scala-cli package foo.sc`, accepts `--js` and `--native`)
- incremental compilation (via Bloop, that relies on zinc)
- compile-only mode (`scala-cli compile foo.sc`, also accepts `-w` / `--watch`, like the other commands)
- Scala version can be specified on the command line (`-S 2.13.7`) or in the scripts themselves (`// using scala 2.13.7` at the top)
- it accepts both `.scala` and `.sc` files
- …

You can quickly run it by using its [`scala-cli.sh`](https://github.com/VirtusLab/scala-cli/raw/master/scala-cli.sh) launcher script for example.

**Comment by alexarchambault** [comment]
I'd love to integrate it in Ammonite somehow. Scala CLI revolves more around "projects" than individual files (a project can be a single `*.sc` or `*.scala` file, but also several of them, or directories containing such files), so that would be more something like `import $project.…`.

Going the other way around, Scala CLI can already run Ammonite on existing projects, like
```text
$ cat a/foo.sc
val n = 2

$ scala-cli repl . --amm
Compiling project (Scala 3.0.2, JVM)
Compiled project (Scala 3.0.2, JVM)
Loading...
Welcome to the Ammonite Repl 2.4.0-23-76673f7f (Scala 3.0.2 Java 11.0.9.1)
@ a.foo.n
res0: Int = 2
```

**Comment by lihaoyi** [comment]
Hah that's an interesting point on the similarities to scala-cli.

However I don't think we're going to be able to drop Ammonite scripts entirely; there are a lot of Ammonite scripts out there. While scala-cli is similar, it feels like there's enough differences for it to not be a drop-in replacement.

I'll see if I have time to do this on the Ammonite codebase. Worst come to worse Ammonite and scala-cli converge, which would probably be a good thing overall

**Comment by lefou** [comment]
_I didn't see this RFC earlier, so I'm adding my 2 cents now._

I pretty much like this approach. It solves the issue, that multiple `$ivy` imports may bring conflicting dependencies into the class path, which we would avoid if we collect and concatenate all import requests and resolve then as a whole.

I can't tell how much work this is, but from a Mill perspective (Mill build scripts are based Ammonite scripts) I'd like to have this paradigm shift rather soon.

**Comment by lefou** [comment]
I think, after we did this, this would also allow us in Mill, to define external module is imported files. Which could be a step forward to provide plugins-as-sources, so they don't need to be pre-compiled.

**Comment by ckipp01** [comment]
Just to throw another opinion in here since it's an RFC, we have heavily heavily relied on Ammonite scripts at one of our clients for my work. Honestly, I can't understate the value that Ammonite has brought to us. We do all of our license reporting, dependency reporting, create graphs for showing interdependencies, etc all with Ammonite. So firstly, thanks for making something so impactful for us and allowing us to truly do _scripting_ with Scala. Ammonite was a trail blazer here.

With that being said, I'm sort of 100% in agreement with what @alexarchambault said about `scala-cli`. We're actually in the process of migrating away from Ammonite because of this. There is better IDE support (Most of our team uses Metals), Scala native support, incremental compilation, and even packaging already there. For us, most of our stuff wasn't a drop-in replacement, but all somewhat trivial to migrate so far. The migration effort just "made sense" because of the punch that scala-cli brings already.

I do hope that if this refactoring takes place, it would offer something that scala-cli doesn't, and I have no idea what that'd be. If not I'm just curious about the effort and the ultimate long-term goal of Ammonite as a script runner while scala-cli is also there as a script runner. Again, talking about this only in terms of script runners.

**Comment by lefou** [comment]
From a Mill perspective, Ammonite is providing features scala-cli currently not  has. Especially the explicit `$ivy` and `$file` imports, which are essential to have a single-source-of-thruth buildfile. But we definitely want all those features scala-cli already has and that are outlined in the RFC :-)

**Comment by lefou** [comment]
Whether we could provided these features on top of scala-cli instead of Ammonite remains to be proven. I for myself didn't had time to look into scala-cli yet.

**Comment by ckipp01** [comment]
> Especially the explicit $ivy and $file imports, which are essential to have a single-source-of-thruth buildfile.

Sure. In `scala-cli` `$ivy` imports are already there and `$file` imports I don't believe are planned since scripts can [automatically reference](https://scala-cli.virtuslab.org/docs/guides/scripts/#difference-with-ammonite-scripts) eachother.

**Comment by lihaoyi** [comment]
Yeah it's a bit of an open question what's the future of Ammonite script running. I see three main options:

1. We keep improving it in Ammonite
2. We abandon it and swap over to scala-cli
3. We abandon it and instead use a thin wrapper around the Mill build tool

It's not immediately obvious to me which is the correct path forward. 

1. Has best compatibility, but would require a lot of feature work (incremental compilation, scala-native, packaging, etc.). It's doable, but it's no doubt a lot of work. And duplicates a lot of work already done in scala-cli and in Mill: this is wasteful, an unnecessary maintenance burden, and offers the potential for diverging behavior on either side.

2. Lets us offload some of the maintenance effort, since scala-cli has its own set of maintainers and developers. But as mentioned is relatively incompatible both feature-wise and interface wise, and pulls in a bunch of dependencies (e.g. bloop) that go against the self-contained nature of the com-lihaoyi ecosystem.

3. Should only require a relatively thin wrapper/preprocessor to turn our scripts into something that Mill understands (e.g. a `mill.scalalib.ScalaModule`), allowing us to re-use the heavily tested and exercised parts of Mill, providing a broad featureset while avoiding re-implementing any major features. It would also possibly provide the smoothest "upgrade" path for scripts that need to be promoted to real projects, since scripts would just be syntactic sugar over a full Mill project, with identical semantics down to the most minute details. Lastly it would preserve the self-contained nature of the com-lihaoyi ecosystem, even improving upon it in some ways (e.g. Mill would no longer strictly need a hard dependency on Ammonite, and could be a totally separate project that happens to use Ammonite for it's `foo.repl` task, assuming we're willing to give up the REPL build mode that AFAICT nobody uses)

**Comment by lefou** [comment]
I especially like the third idea. Mill already has anything to compile build scripts. I think handling of $ivy and $file import might be straighforward and should work as described in this very RFC. Some features need to be ported from Ammonite to Mill though, e.g. watch support. But given that both projects share their dependencies and environments, this might be not that hard, too.
That being said, it's still a tough task. But we have a very good test suite.

Sorry for my focus on Mill here, but I rarely used Ammonite standalone.

**Comment by thinkiny** [comment]
> Yeah it's a bit of an open question what's the future of Ammonite script running. I see three main options:
> 
>     1. We keep improving it in Ammonite
> 
>     2. We abandon it and swap over to scala-cli
> 
>     3. We abandon it and instead use a thin wrapper around the Mill build tool
> 
> 
> It's not immediately obvious to me which is the correct path forward.
> 
>     1. Has best compatibility, but would require a lot of feature work (incremental compilation, scala-native, packaging, etc.). It's doable, but it's no doubt a lot of work. And duplicates a lot of work already done in scala-cli and in Mill: this is wasteful, an unnecessary maintenance burden, and offers the potential for diverging behavior on either side.
> 
>     2. Lets us offload some of the maintenance effort, since scala-cli has its own set of maintainers and developers. But as mentioned is relatively incompatible both feature-wise and interface wise, and pulls in a bunch of dependencies (e.g. bloop) that go against the self-contained nature of the com-lihaoyi ecosystem.
> 
>     3. Should only require a relatively thin wrapper/preprocessor to turn our scripts into something that Mill understands (e.g. a `mill.scalalib.ScalaModule`), allowing us to re-use the heavily tested and exercised parts of Mill, providing a broad featureset while avoiding re-implementing any major features. It would also possibly provide the smoothest "upgrade" path for scripts that need to be promoted to real projects, since scripts would just be syntactic sugar over a full Mill project, with identical semantics down to the most minute details. Lastly it would preserve the self-contained nature of the com-lihaoyi ecosystem, even improving upon it in some ways (e.g. Mill would no longer strictly need a hard dependency on Ammonite, and could be a totally separate self-contained project that happens to use Ammonite for it's `foo.repl` task, assuming we're willing to give up the REPL build mode that AFAICT nobody uses). It won't be as compatible as (1.), but could be made more compatible than (2.)

I vote for the third idea. scala-cli seems good at replacing ammonite, but mill gives more than that

**Comment by nafg** [comment]
>  give up the REPL build mode that AFAICT nobody uses

You mean `mill --repl`? I think that's an important feature. But why would it have to go away?

**Comment by nafg** [comment]
>  But as mentioned is relatively incompatible both feature-wise and interface wise,

Can you elaborate on this? I'm a little confused actually. https://scala-cli.virtuslab.org/docs/guides/scripts/#difference-with-ammonite-scripts says

> scala-cli requires all scripts to be passed beforehand, either one-by-one, or by putting them in a directory, and passing the directory to scala-cli

but https://scala-cli.virtuslab.org/docs/commands/run#define-source-files-in-using-directives says

> You can also add source files with the using directive //> using file:


> and pulls in a bunch of dependencies (e.g. bloop) that go against the self-contained nature of the com-lihaoyi ecosystem.

How important is this?

**Comment by lefou** [comment]
Hello everyone. I just wanted to inform you, that Mill is currently in good progress to compile and run it's build scripts in a self-contained and Ammonite-independend way, thanks to @lihaoyi .

* https://github.com/com-lihaoyi/mill/pull/2377
